numpy/f2py/crackfortran.py

```diff
@@ -2954,9 +2991,155 @@ def analyzevars(block):
                 del vars[n]
     return vars
 
+
 analyzeargs_re_1 = re.compile(r'\A[a-z]+[\w$]*\Z', re.I)
 
 
+def param_eval(v, g_params, params, dimspec=None):
+    """
+    Creates a dictionary of indices and values for each parameter in a
+    parameter array to be evaluated later.
+
+    WARNING: It is not possible to initialize multidimensional array
+    parameters e.g. dimension(-3:1, 4, 3:5) at this point. This is because in
+    Fortran initialization through array constructor requires the RESHAPE
+    intrinsic function. Since the right-hand side of the parameter declaration
+    is not executed in f2py, but rather at the compiled c/fortran extension,
+    later, it is not possible to execute a reshape of a parameter array.
+    One issue remains: if the user wants to access the array parameter from
+    python, we should either
+    1) allow them to access the parameter array using python standard indexing
+       (which is often incompatible with the original fortran indexing)
+    2) allow the parameter array to be accessed in python as a dictionary with
+       fortran indices as keys
+    We are choosing 2 for now.
+    """
+    if dimspec is None:
+        try:
+            p = eval(v, g_params, params)
+        except Exception as msg:
+            p = v
+            outmess('param_eval: got "%s" on %s\n' % (msg, repr(v)))
+        return p
+
+    # This is an array parameter.
+    # First, we parse the dimension information
+    if len(dimspec) < 2 or dimspec[::len(dimspec)-1] != "()":
+        raise ValueError(f'param_eval: dimension {dimspec} can\'t be parsed')
+    dimrange = dimspec[1:-1].split(',')
+    if len(dimrange) == 1:
+        # e.g. dimension(2) or dimension(-1:1)
+        dimrange = dimrange[0].split(':')
+        # now, dimrange is a list of 1 or 2 elements
+        if len(dimrange) == 1:
+            bound = param_parse(dimrange[0], params)
+            dimrange = range(1, int(bound)+1)
+        else:
+            lbound = param_parse(dimrange[0], params)
+            ubound = param_parse(dimrange[1], params)
+            dimrange = range(int(lbound), int(ubound)+1)
+    else:
+        outmess(f'param_eval: multidimensional array parameters ({dimspec}) '
+                'not supported\n')
```

Comment:
I don't understand this line; we print a message claiming not to support it, but the function keeps running anyway. So either this should raise an exception, or we should adjust the message to indicate it's only _partially_ supported.

Reply:
This was done in response to [@pearu's comment](https://github.com/numpy/numpy/pull/15457#discussion_r445838791):

> I would just report this via outmess rather than raising an exception because the corresponding case is often irrelevant for the f2py task.

Reply:
I think it's better to raise the exception - if you do that, then an appropriate place up the stack will emit a suitable error message, and you don't have to guess what the rest of the function does after this message is printed.

Reply:
It's consistent with what we do on line 3031

Reply:
I disagree with the need to raise an exception: the aim of f2py is to create wrapper functions for Fortran routines and this requires only partial interpretation of Fortran code that may or may not be relevant for creating wrapper functions. Most of the Fortran code is irrelevant for creating the wrapper functions and if one raises an exception on every piece of Fortran code that f2py is not able to interpret, as is the given case here, you basically break f2py.

Recall that f2py does not aim at wrapping arbitrary Fortran code to Python without any user intervention, it would not just be possible in general. According to the recommended usage of f2py, one should use signature files to design the UI and functionality of wrapper functions. Only rarely one is able to apply f2py directly to Fortran source codes and get satisfactory wrapper functions. f2py can be used to generate initial versions of signature files that users ought to modify to their actual needs. If f2py would raise an exception on every Fortran codelet that it does not support yet, one would not be able to generate the signature files. Hence, your suggestion to raise an exception here breaks the normal usage of f2py.

A rule of thumb: in f2py crackfortran.py file, one must not raise exceptions on branches that correspond to f2py not supporting a particular feature in Fortran: such branches are often irrelevant, and printing out a message that a branch of unsupported Fortran construct is met, is sufficient to better design the signature file or submit a feature request for f2py to automatically support the particular Fortran constructs.

Raising an exception in the above-described situations is contrary to f2py model and all such proposals must be rejected.

Reply:
However, this is an exception to the above-mentioned rule: when the inputs to the f2py reader are a signature file or an f2py-directive, raising an exception on an unsupported Fortran construct is ok.

Reply:
> and if one raises an exception on every piece of Fortran code that f2py is not able to interpret, as is the given case here, you basically break f2py.

You only break f2py if the exception propagates all the way to the caller, which it does not here. It's far easier to reason about functions that raise exceptions when their preconditions are not met and they are unable to proceed - the caller is still free to catch the exception and recover, which is exactly what is happening here.

To emphasize that point - with an exception here, we can add a test for this and verify that it fails as we expect. With a message here, I expect it throws an exception anyway, but it's now a nonsense one cause by continuing with invalid preconditions.

I understand that the goal of f2py is to in many places perform a best effort parsing - but that doesn't mean that every helper function should perform a best effort.

Reply:
With https://github.com/numpy/numpy/blob/1adc24adcdc1dcc1d52f8eba2065b6f243980783/numpy/f2py/crackfortran.py#L2596-L2599 I agree with @eric-wieser to raise an exception here or elsewhere as long as one keeps in mind the f2py model: f2py should not fail on Fortran constructs that support has not been implemented.